RandBLAS/base.hh

```diff
@@ -1,131 +1,191 @@
+    RNGState(typename RNG::ukey_type const& k) : counter{{}}, key(k) {}
 
-    RNGState(const RNGState &s);
-    RNGState(RNGState &&s);
+    /// construct from an initial counter and key
+    RNGState(typename RNG::ctr_type const& c, typename RNG::key_type const& k) : counter(c), key(k) {}
 
-    RNGState(uint32_t c0, uint32_t k0);
+    /// move construct from an initial counter and key
+    RNGState(typename RNG::ctr_type &&c, typename RNG::key_type &&k) : counter(std::move(c)), key(std::move(k)) {}
 
-    template <typename T_state>
-    RNGState(const T_state &in_state);
+    /// @name conversions
+    ///{
+    RNGState(std::tuple<typename RNG::ctr_type, typename RNG::key_type> const& tup);
+    RNGState(std::tuple<typename RNG::ctr_type, typename RNG::key_type> && tup);
+    operator std::tuple<typename RNG::ctr_type const&, typename RNG::key_type const&> () const;
+    operator std::tuple<typename RNG::ctr_type&, typename RNG::key_type&> ();
+    ///}
 
-    RNGState &operator=(const RNGState &s);
-    RNGState &operator=(RNGState &&s);
+    typename RNG::ctr_type counter; ///< the counter
+    typename RNG::key_type key;     ///< the key
 };
 
-std::ostream &operator<<(std::ostream &out, const RNGState &s);
-
-template <typename T_gen>
-struct Random123_RNGState {
-
-    typedef T_gen gen_type;
-    typedef typename T_gen::key_type key_type;
-    typedef typename T_gen::ctr_type ctr_type;
-
-    ctr_type ctr{};
-    key_type key{};
-
-    const int len_c = ctr_type::static_size;
-    const int len_k = key_type::static_size;
-
-    Random123_RNGState(const RNGState &s);
-};
+template <typename RNG>
+RNGState<RNG>::RNGState(
+    std::tuple<typename RNG::ctr_type,
+    typename RNG::key_type> const& tup
+) : counter(std::get<0>(tup)),
+    key(std::get<1>(tup)) {}
+
+template <typename RNG>
+RNGState<RNG>::RNGState(
+    std::tuple<typename RNG::ctr_type,
+    typename RNG::key_type> && tup
+) : counter(std::move(std::get<0>(tup))),
+    key(std::move(std::get<1>(tup))) {}
+
+template <typename RNG>
+RNGState<RNG>::operator std::tuple<typename RNG::ctr_type const&, typename RNG::key_type const&> () const {
+    return std::tie(std::as_const(counter), std::as_const(key));
+}
 
-// Convert from Random123_RNGState to RNGState
-template <typename T_state>
-RNGState::RNGState(
-    const T_state &in_state
-) : len_c(in_state.len_c),
-    len_k(in_state.len_k)
-{
-    typedef typename T_state::gen_type gen_type;
-    Philox ph;
-    Threefry tf;
-    gen_type gt;
-    auto gtid = ::std::type_index(typeid(gt));
-    if (gtid == ::std::type_index(typeid(ph))) {
-        this->rng_name = RNGName::Philox;
-    } else if (gtid == ::std::type_index(typeid(tf))) {
-        this->rng_name = RNGName::Threefry;
-    } else {
-        throw std::runtime_error(std::string("Unknown gen_type."));
-    }
+template <typename RNG>
+RNGState<RNG>::operator std::tuple<typename RNG::ctr_type&, typename RNG::key_type&> () {
+    return std::tie(counter, key);
+}
```

Comment:
What are these functions supposed to do?

Reply:
those are conversion functions so RNGState can be converted to a std::tuple, which is useful with structured bindings

Reply:
update: I'll remove these since they're not used any ware, and structured bindings work out of the box with classes/structs that have all public non static member variables.

Reply:
resolved in c6077f65